Source/WebCore/platform/haiku/PlatformKeyboardEventHaiku.cpp

```diff
@@ -324,6 +324,332 @@ static int windowsKeyCodeForKeyEvent(char singleByte, int keyCode)
+{
+    switch (singleByte) {
+        
+        case B_FUNCTION_KEY:
+            switch (keyCode) {
+                case B_F1_KEY:
+                    return "F1";
+                case B_F2_KEY:
+                    return "F2";
+                case B_F3_KEY:
+                    return "F3";
+                case B_F4_KEY:
+                    return "F4";
+                case B_F5_KEY:
+                    return "F5";
+                case B_F6_KEY:
+                    return "F6";
+                case B_F7_KEY:
+                    return "F7";
+                case B_F8_KEY:
+                    return "F8";
+                case B_F9_KEY:
+                    return "F9";
+                case B_F10_KEY:
+                    return "F10";
+                case B_F11_KEY:
+                    return "F11";
+                case B_F12_KEY:
+                    return "F12";
+                case B_PRINT_KEY:
+                    return "Print";
+                case B_PAUSE_KEY:
+                    return "Pause";
+                case B_SCROLL_KEY:
+                    return "ScrollLock";
+                case B_SHIFT_KEY:
+                    return "Shift";
+                case B_LEFT_CONTROL_KEY: // B_CONTROL_KEY = B_F3_KEY..?
+                case B_RIGHT_CONTROL_KEY:
+                    return "Control";
+            }
+            
+            break;
+            
+                case B_BACKSPACE:
+                    return "Backspace";
+                case B_LEFT_ARROW:
+                    return "ArrowLeft";
+                case B_RIGHT_ARROW:
+                    return "ArrowRight";
+                case B_UP_ARROW:
+                    return "ArrowUp";
+                case B_DOWN_ARROW:
+                    return "ArrowDown";
+                case B_INSERT:
+                    return "Insert";
+                case B_ENTER:
+                    return "Enter";
+                case B_DELETE:
+                    return "Delete";
+                case B_HOME:
+                    return "Home";
+                case B_END:
+                    return "End";
+                case B_PAGE_UP:
+                    return "PageUp";
+                case B_PAGE_DOWN:
+                    return "PageDown";
+                case B_TAB:
+                    return "Tab";
+                case B_SPACE:
+                    return " "; // (20) SPACEBAR
+                    
+                    
+                case B_COMMAND_KEY:
+                    return "Alt";
+                    
+                case B_OPTION_KEY:
+                    return "Compose"; // ?? why compose? fx also said "Process"
+                case B_MENU_KEY:
+                    return "ContextMenu";
+                case B_ESCAPE:
+                    return "Escape";
+                    
+                default: {
+                    String keyVal = String::format("%c", singleByte);
```

Comment:
this will not work for non-ASCII characters. The bytes are an utf-8 sequence, and I'd say they should be fully preserved, not just the first byte.

Reply:
Thanks for your feedback! I've attempted better work in PlatformKeyboard just now and will look at moving those features.